# Déjame escuchar música

# Contenido <a id='back'></a>

* [Introducción](#intro)
* [Etapa 1. Descripción de los datos](#data_review)
    * [Conclusiones](#data_review_conclusions)
* [Etapa 2. Preprocesamiento de datos](#data_preprocessing)
    * [2.1 Estilo del encabezado](#header_style)
    * [2.2 Valores ausentes](#missing_values)
    * [2.3 Duplicados](#duplicates)
    * [2.4 Conclusiones](#data_preprocessing_conclusions)
* [Etapa 3. Prueba de hipótesis](#hypothesis)
    * [3.1 Hipótesis 1: actividad de los usuarios y las usuarias en las dos ciudades](#activity)
* [Conclusiones](#end)

## Introducción <a id='intro'></a>
Como analista de datos, tu trabajo consiste en analizar datos para extraer información valiosa de ellos y tomar decisiones en consecuencia. Esto implica pasar por diferentes etapas, como la descripción general de los datos, el preprocesamiento y la prueba de hipótesis.

Siempre que investiguemos, necesitamos formular hipótesis que después podamos probar. A veces, aceptaremos estas hipótesis; otras veces, las rechazaremos. Para tomar las decisiones adecuadas, una empresa debe ser capaz de entender si está haciendo las suposiciones correctas.

En este proyecto, compararás las preferencias musicales de las ciudades de Springfield y Shelbyville. Estudiarás datos reales de música online para probar la hipótesis que planteamos a continuación y compararás el comportamiento de los usuarios de estas dos ciudades.

### Objetivo:
Probar la hipótesis:
1. La actividad de los usuarios difiere según el día de la semana y dependiendo de la ciudad.


### Etapas
Los datos del comportamiento de los usuarios se almacenan en el archivo `/datasets/music_project_en.csv`. No hay información sobre la calidad de los datos, así que necesitarás examinarlos antes de probar la hipótesis.

Primero, debes evaluar la calidad de los datos y ver si los problemas son significativos. Más tarde, durante el preprocesamiento de datos, deberás abordar los problemas más críticos.

Tu proyecto contará con estas tres etapas:
 1. Descripción de los datos.
 2. Preprocesamiento de los datos.
 3. Prueba de la hipótesis.








[Volver a Contenidos](#back)

## Etapa 1. Descripción de los datos <a id='data_review'></a>

Abre los datos y examínalos.

Etapa 1.1. Necesitarás `pandas`, así que impórtalo.

In [18]:
# Importa pandas
import pandas as pd

Etapa 1.2. Lee el archivo `music_project_en.csv` de la carpeta `/datasets/` y guárdalo en la variable `df`:

In [19]:
# Lee el archivo y almacénalo en df
df = pd.read_csv('DATASETS/music_project_en.csv')

Etapa 1.3. Muestra las 10 primeras filas de la tabla:

In [20]:
# Obtén las 10 primeras filas de la tabla df
print(df.head(10))

     userID                        Track            artist   genre  \
0  FFB692EC            Kamigata To Boots  The Mass Missile    rock   
1  55204538  Delayed Because of Accident  Andreas Rönnberg    rock   
2    20EC38            Funiculì funiculà       Mario Lanza     pop   
3  A3DD03C9        Dragons in the Sunset        Fire + Ice    folk   
4  E2DC1FAE                  Soul People        Space Echo   dance   
5  842029A1                       Chains          Obladaet  rusrap   
6  4CB90AA5                         True      Roman Messer   dance   
7  F03E1C1F             Feeling This Way   Polina Griffith   dance   
8  8FA1D3BE                     L’estate       Julia Dalia  ruspop   
9  E772D5C0                    Pessimist               NaN   dance   

        City        time        Day  
0  Shelbyville  20:28:33  Wednesday  
1  Springfield  14:07:09     Friday  
2  Shelbyville  20:58:07  Wednesday  
3  Shelbyville  08:37:09     Monday  
4  Springfield  08:34:34     Monday  
5

Etapa 1.4. Obtén la información general sobre la tabla con el método info().

In [21]:
# Obtén la información general sobre nuestros datos
df.info()

# Obtenemos el número de filas duplicadas
df.duplicated().sum()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65079 entries, 0 to 65078
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0     userID  65079 non-null  object
 1   Track     63736 non-null  object
 2   artist    57512 non-null  object
 3   genre     63881 non-null  object
 4     City    65079 non-null  object
 5   time      65079 non-null  object
 6   Day       65079 non-null  object
dtypes: object(7)
memory usage: 3.5+ MB


3826

Estas son nuestras observaciones sobre la tabla. Contiene siete columnas que almacenan los mismos tipos de datos: `object`.

Según la documentación:
- `' userID'`: identificador del usuario;
- `'Track'`: título de la canción;
- `'artist'`: nombre del artista;
- `'genre'`: género de la canción;
- `'City'`: ciudad del usuario;
- `'time'`: la hora exacta en la que se reprodujo la canción;
- `'Day'`: día de la semana.

Podemos ver tres problemas con el estilo en los encabezados de la tabla:
1. Algunos encabezados están en mayúsculas, otros en minúsculas.
2. Hay espacios en algunos encabezados.
3. No se usa "_" para separar dos palabras (nomenclatura snake case).




### Escribe algunas observaciones por tu parte. Contesta a las siguientes preguntas: <a id='data_review_conclusions'></a>

`1.   ¿Qué tipo de datos hay en las filas? ¿Cómo podemos saber qué almacenan las columnas?`

`2.   ¿Hay suficientes datos para proporcionar respuestas a nuestra hipótesis o necesitamos más información?`

`3.   ¿Notaste algún problema en los datos, como valores ausentes, duplicados o tipos de datos incorrectos?`

Escribe aquí tus respuestas:

1. En las filas hay datos de tipo cadena o string según la información que nos dió el método info(), para saber el tipo de dato se usa el metodo info() en el dataframe y si se desea una visualización de algunos datos se puede usar el método head() para ver las primeras filas o sample() para obtener una muestra aleatoria del dataframe

2. Hay suficientes datos para proporcionar una respuesta, vemos que se tiene algunos datos ausentes en las columnas de 'Track', 'artist', 'genre'; pero no existe un vuen volumen de información no nula.

3. En cuanto a los datos, existen valores ausentes en 3 columnas como lo comentamos en el punto 2, las columnas 'Track', 'artist' y 'genre' cuentan con valores ausentes: 1343, 7567, 1198 respectivamente; con resepecto a duplicados, según el metodo "duplicated().sum()", aparentemente tenemos 3826 filas duplicadas; en cuanto a datos incorrectos se podría decir que la columna 'time' tiene datos incorrectos, ya que debería transormarse a un dato tipo hora si es que se necesitara para realizar algún cálculo, pero como no es así se puede dejar como object o string.

[Volver a Contenidos](#back)

## Etapa 2. Preprocesamiento de los datos <a id='data_preprocessing'></a>

Tu objetivo aquí es preparar los datos para analizarlos.
El primer paso es resolver los problemas con los encabezados. Después podemos avanzar a los valores ausentes y duplicados. ¡Empecemos!

Vamos a corregir el formato en los encabezados de la tabla.


### Estilo del encabezado <a id='header_style'></a>
Etapa 2.1. Muestra los encabezados de la tabla (los nombres de las columnas):

In [22]:
# Muestra los nombres de las columnas
df.columns

Index(['  userID', 'Track', 'artist', 'genre', '  City  ', 'time', 'Day'], dtype='object')

Vamos cambiar los encabezados de la tabla siguiendo las reglas estilísticas convencionales:
*   Todos los caracteres deben ser minúsculas.
*   Elimina los espacios.
*   Si el nombre tiene varias palabras, utiliza snake_case.


Anteriormente, aprendiste una forma automática de cambiar el nombre de las columnas. Vamos a aplicarla ahora.

Etapa 2.2. Utiliza el bucle for para iterar sobre los nombres de las columnas y poner todos los caracteres en minúsculas. Cuando hayas terminado, vuelve a mostrar los encabezados de la tabla:

In [23]:
# Bucle en los encabezados que lo pone todo en minúsculas
new_col_names = []

for old_name in df.columns:
    lowercase_name = old_name.lower()
    new_col_names.append(lowercase_name)

df.columns = new_col_names

Etapa 2.3. Ahora, utilizando el mismo método, elimina los espacios al principio y al final de los nombres de las columnas y muestra los nombres de las columnas de nuevo:

In [24]:
# Bucle en los encabezados que elimina los espacios
new_col_names = []

for old_name in df.columns:
    lowercase_name = old_name.strip()
    new_col_names.append(lowercase_name)

df.columns = new_col_names

Etapa 2.4. Necesitamos aplicar la regla de snake_case en la columna `userid`. Debe ser `user_id`. Cambia el nombre de esta columna y muestra los nombres de todas las columnas cuando hayas terminado.

In [25]:
# Cambia el nombre de la columna "userid"
new_column = {
    "userid" : "user_id"
}
df.rename(columns = new_column, inplace=True)

df.columns

Index(['user_id', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='object')

Etapa 2.5. Comprueba el resultado. Muestra los encabezados una vez más:

In [26]:
# Comprueba el resultado: lista de encabezados
df.columns

Index(['user_id', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='object')

[Volver a Contenidos](#back)

### Valores ausentes <a id='missing_values'></a>
 Etapa 2.5. Primero, encuentra el número de valores ausentes en la tabla. Debes utilizar dos métodos para obtener el número de valores ausentes.

In [27]:
# Calcula el número de valores ausentes
df.isna().sum()

user_id       0
track      1343
artist     7567
genre      1198
city          0
time          0
day           0
dtype: int64

No todos los valores ausentes afectan de la misma forma a la investigación. Por ejemplo, los valores ausentes en 'track' y 'artist' no son cruciales para el análisis, ya que estos datos son más descriptivos que analíticos. Por eso, puedes reemplazarlos directamente con un valor predeterminado como el string 'unknown' (desconocido).

En cambio, los valores ausentes en 'genre' sí pueden influir en la comparación entre las preferencias musicales de Springfield y Shelbyville. En un escenario real, lo ideal sería investigar por qué faltan estos datos e intentar recuperarlos. Pero en este proyecto no tenemos esa posibilidad, así que deberás rellenar esos valores con un valor predeterminado.


Como vimos anteriormente en las lecciones, la mejor manera de hacerlo es crear una lista con los nombres de las columnas que necesitan reemplazarse.  Luego, utilizar esta lista para iterar sobre cada columna y realizar el reemplazo correspondiente.

Etapa 2.6. Sustituye los valores ausentes en las columnas `'track'`, `'artist'` y `'genre'` con el string `'unknown'`.

1. Crea una lista llamada columns_to_replace que contenga los nombres de las columnas 'track', 'artist' y 'genre'.

2. Usa un bucle for para iterar sobre cada columna en columns_to_replace.

3. Dentro del bucle, sustituye los valores ausentes en cada columna con el string `'unknown'`.

In [28]:
# Bucle en los encabezados reemplazando los valores ausentes con 'unknown'
columns_to_replace = ['track', 'artist', 'genre']

for column in columns_to_replace:
    df[column].fillna('unknown', inplace=True)


/var/folders/05/s3zd8b1n5qsd1bpkp2xg179w0000gn/T/ipykernel_2191/2758327399.py:5: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df[column].fillna('unknown', inplace=True)


Etapa 2.7. Ahora comprueba el resultado para asegurarte de que no falten valores ausentes por reemplazar en el conjunto de datos. Para ello, cuenta los valores ausentes una vez más.

In [29]:
# Cuenta los valores ausentes
df.isna().sum()

user_id    0
track      0
artist     0
genre      0
city       0
time       0
day        0
dtype: int64

[Volver a Contenidos](#back)

### Duplicados <a id='duplicates'></a>
Etapa 2.8. Encuentra el número de duplicados explícitos en la tabla. Una vez más, debes aplicar dos métodos para obtener la cantidad de duplicados explícitos.

In [30]:
# Cuenta los duplicados explícitos
df.duplicated().sum()

3826

Etapa 2.9. Ahora, elimina todos los duplicados. Para ello, llama al método que hace exactamente esto.

In [31]:
# Elimina los duplicados explícitos
df.drop_duplicates(inplace=True)
df.drop_duplicates().reset_index(drop=True)

user_id                              track            artist  \
0      FFB692EC                  Kamigata To Boots  The Mass Missile   
1      55204538        Delayed Because of Accident  Andreas Rönnberg   
2        20EC38                  Funiculì funiculà       Mario Lanza   
3      A3DD03C9              Dragons in the Sunset        Fire + Ice   
4      E2DC1FAE                        Soul People        Space Echo   
...         ...                                ...               ...   
61248  729CBB09                            My Name            McLean   
61249  D08D4A55  Maybe One Day (feat. Black Spade)       Blu & Exile   
61250  C5E3A0D5                          Jalopiina           unknown   
61251  321D0506                      Freight Train     Chas McDevitt   
61252  3A64EF84          Tell Me Sweet Little Lies      Monica Lopez   

            genre         city      time        day  
0            rock  Shelbyville  20:28:33  Wednesday  
1            rock  Springfield  14:07:09     Friday  
2             pop  Shelbyville  20:58:07  Wednesday  
3            folk  Shelbyville  08:37:09     Monday  
4           dance  Springfield  08:34:34     Monday  
...           ...          ...       ...        ...  
61248         rnb  Springfield  13:32:28  Wednesday  
61249         hip  Shelbyville  10:00:00     Monday  
61250  industrial  Springfield  20:09:26     Friday  
61251        rock  Springfield  21:43:59     Friday  
61252     country  Springfield  21:59:46     Friday  

[61253 rows x 7 columns]

Etapa 2.10. Comprobemos ahora si conseguimos eliminar todos los duplicados. Cuenta los duplicados explícitos una vez más para asegurarte de haberlos eliminado todos:

In [32]:
# Comprueba de nuevo si hay duplicados
df.duplicated().sum()

0

Ahora queremos deshacernos de los duplicados implícitos en la columna `genre`. Por ejemplo, el nombre de un género se puede escribir de varias formas. Dichos errores también pueden afectar al resultado.

Etapa 2.11. Primero debemos mostrar una lista de nombres de géneros únicos, por orden alfabético. Para ello:
1. Extrae la columna `genre` del DataFrame.
2. Llama al método que devolverá todos los valores únicos en la columna extraída.


In [33]:
# Inspecciona los nombres de géneros únicos
df['genre'].unique()

array(['rock', 'pop', 'folk', 'dance', 'rusrap', 'ruspop', 'world',
       'electronic', 'unknown', 'alternative', 'children', 'rnb', 'hip',
       'jazz', 'postrock', 'latin', 'classical', 'metal', 'reggae',
       'triphop', 'blues', 'instrumental', 'rusrock', 'dnb', 'türk',
       'post', 'country', 'psychedelic', 'conjazz', 'indie',
       'posthardcore', 'local', 'avantgarde', 'punk', 'videogame',
       'techno', 'house', 'christmas', 'melodic', 'caucasian',
       'reggaeton', 'soundtrack', 'singer', 'ska', 'salsa', 'ambient',
       'film', 'western', 'rap', 'beats', "hard'n'heavy", 'progmetal',
       'minimal', 'tropical', 'contemporary', 'new', 'soul', 'holiday',
       'german', 'jpop', 'spiritual', 'urban', 'gospel', 'nujazz',
       'folkmetal', 'trance', 'miscellaneous', 'anime', 'hardcore',
       'progressive', 'korean', 'numetal', 'vocal', 'estrada', 'tango',
       'loungeelectronic', 'classicmetal', 'dubstep', 'club', 'deep',
       'southern', 'black', 'folkrock', 

Vamos a examinar la lista para identificar **duplicados implícitos** del género `hiphop`, es decir, nombres mal escritos o variantes que hacen referencia al mismo género musical.

Algunos de los duplicados que encontrarás son:

* `hip`  
* `hop`  
* `hip-hop`  

Para solucionarlo, vamos a crear una función llamada `replace_wrong_genres()` que tendrá dos parámetros:

* `wrong_genres`: una lista con todos los valores que deben ser reemplazados.  
* `correct_genre`: un string que se utilizará como valor de reemplazo.

El objetivo de esta función es **corregir los valores en la columna `'genre'` del DataFrame `df`**, reemplazando cada valor de la lista `wrong_genres` por `correct_genre`.

Etapa 2.12.
1. Define una función llamada `replace_wrong_genres()` que reciba dos parámetros: `wrong_genres` y `correct_genre`.

2. Dentro de la función, utiliza un bucle `for` para iterar sobre cada valor en la lista `wrong_genres`.

3. En cada iteración, accede a la columna `'genre'` del DataFrame `df` y utiliza el método `.replace()` para sustituir el valor incorrecto por `correct_genre`.

4. Llama a la función y pasa como argumentos:
   - Una lista con los duplicados implícitos: `['hip', 'hop', 'hip-hop']`
   - El string de reemplazo: `'hiphop'`

In [34]:
# Función para reemplazar los duplicados implícitos
def replace_wrong_genres(wrong_genres, correct_genre):
    for wrong_genre in wrong_genres:
        df['genre'] = df['genre'].replace(wrong_genre, correct_genre)
    return df['genre']

Etapa 2.13. Ahora, llama a `replace_wrong_genres()` y pásale estos argumentos para que retire los duplicados implícitos (`hip`, `hop` y `hip-hop`) y los sustituya por `hiphop`:

In [35]:
# Elimina los duplicados implícitos
replace_wrong_genres(['hip', 'hop', 'hip-hop'], 'hiphop')

0              rock
1              rock
2               pop
3              folk
4             dance
            ...    
65074           rnb
65075        hiphop
65076    industrial
65077          rock
65078       country
Name: genre, Length: 61253, dtype: object

Etapa 2.14. Asegúrate de que los nombres duplicados se hayan eliminado. Muestra la lista de valores únicos de la columna `'genre'` una vez más:

In [36]:
# Comprueba de nuevo los duplicados implícitos

#Comprobamos si existe algún dato que fue sustituido
print('Filas que contenga en genero erroneo de hiphop:')
print(df[(df['genre'] == 'hip') | (df['genre'] == 'hop') | (df['genre'] == 'hip-hop')], '\n')

#Comprobamos si existen valores en 'genre' con 'hiphop'
print('Filas con género correcto (hiphop)')
print(df.query("genre == 'hiphop'"), '\n')

print('Valores únicos:\n', df['genre'].unique())

Filas que contenga en genero erroneo de hiphop:
Empty DataFrame
Columns: [user_id, track, artist, genre, city, time, day]
Index: [] 

Filas con género correcto (hiphop)
        user_id                                track  \
20     201CF2A8                     Ya'll In Trouble   
46     825997A5                     Glorious Feeling   
79     1DA07AA4                              Cardi B   
81     CCD46819                               Narcos   
82       80396F                           Ay Caramba   
...         ...                                  ...   
65034  C1C4C94D                              Payback   
65041  78713DBB          Feels Like Saving The World   
65043  CDCAFD62                               Payday   
65052  72CB2D36  Yeezy Taught Her (feat. King Louie)   
65075  D08D4A55    Maybe One Day (feat. Black Spade)   

                                                  artist   genre         city  \
20     Lil Tee Chill Tank Young Buck Brother Mohammed...  hiphop  Springfield

[Volver a Contenidos](#back)

### Observaciones <a id='data_preprocessing_conclusions'></a>

`Redacta un breve resumen de lo que descubriste al analizar los datos. Tu respuesta debe identificar los problemas detectados, explicar cómo los resolviste y describir cómo esas acciones mejoran la calidad del análisis.`

Etapa 2.14. Hasta el momento se detectó que el nombre de las cabeceras del dataframe estaba mal nombradas. Por lo cual se procedió a renombrarlas colocando todas las letras en minúscula, borrando espacios en blanco al inicio y final y se usó la nomenclatura snake case en el caso de la cabecera "user_id".
También se detectó valores ausentes en las columnas de "track", "artist" y "genre", por lo cual se procesió a reemplazar los valores ausentes con 'unknown' de las 3 columnas.
Luego, se detectaron filas duplicadas, por lo cual se eliminó dichas filas y se reseteó los indices para que sean correlativos en el dataframe.
Por último, se detectó duplicados explícitos en la columna 'genre', por ejemplo para hiphop se observó 3 tipos adicionales que significa lo mismo. Por lo cual se procedió a reemplazar los valores incorrectos con el valor correcto.
Con todas las acciones realizadas la data se encuentra más limpia, con menos errores y se puede obtener un análisis más exacto para la aceptación o rechazo de la hipótesis.

[Volver a Contenidos](#back)

## Etapa 3. Prueba de la hipótesis <a id='hypothesis'></a>

### Tarea: Comparar el comportamiento de los usuarios en las dos ciudades <a id='activity'></a>

La hipótesis que queremos probar plantea que existen diferencias en la forma en que los usuarios de Springfield y Shelbyville consumen música.
Para analizar esto, nos enfocaremos en los datos correspondientes a tres días específicos de la semana: lunes, miércoles y viernes.

El análisis consistirá en comparar la cantidad de canciones reproducidas por los usuarios de cada ciudad en esos días. Esto nos permitirá observar posibles patrones o diferencias en los hábitos de consumo entre Springfield y Shelbyville.

Para llevar a cabo este análisis, es importante seguir el enfoque de dividir-aplicar-combinar, del que ya hablamos en la lección. En este caso:

*  Dividir: separamos los datos en grupos según la ciudad.

*  Aplicar: dentro de cada grupo, contamos cuántas canciones se reprodujeron.

*  Combinar: reunimos los resultados en una estructura que nos permita comparar fácilmente ambas ciudades.

Este procedimiento debe repetirse de forma independiente para cada uno de los tres días seleccionados. El resultado final debe mostrar el número de reproducciones por ciudad en cada uno de esos días.

Una posible forma de estructurar el código sería con la siguiente expresión:

`df.groupby(by='...')['...'].method()`

Deberás completar los argumentos correspondientes para agrupar por ciudad y contar las canciones reproducidas. Este enfoque te dará una visión clara y comparativa del comportamiento de los usuarios en ambas ciudades.

Etapa 3.1. Cuenta cuántas canciones se reprodujeron en cada ciudad utilizando la columna 'track' como referencia.

In [37]:
# Cuenta las canciones reproducidas en cada ciudad
df.groupby('city')
df.groupby('city')['track'].count().sort_values(ascending=False)

city
Springfield    42741
Shelbyville    18512
Name: track, dtype: int64

`Redacta brevemente tus observaciones sobre los resultados. ¿Qué diferencias notaste entre Springfield y Shelbyville? ¿A qué podrían deberse esas diferencias?`

Etapa 3.2. Se puede observar que en la ciudad de 'Springfield' se reprodujeron más del doble de canciones que en la ciudad 'Shelbyville'. Parece que en la ciudad de 'Springfield' les gusta escuchar mucho música mientras realizan sus labores.

Etapa 3.3.
1. Agrupa los datos por día de la semana y cuenta cuántas canciones se reprodujeron los lunes, miércoles y viernes.

2. Utiliza el mismo método de conteo que antes, pero ahora cambia la columna de agrupación para enfocarte en el día.

3. Esto te permitirá identificar posibles patrones de escucha según el día de la semana.


In [38]:
# Calcula las canciones reproducidas en cada uno de los tres días
df.groupby('day')['track'].count().sort_values(ascending=False)

day
Friday       21840
Monday       21354
Wednesday    18059
Name: track, dtype: int64

`Describe brevemente qué observaste al comparar los lunes, miércoles y viernes. ¿Hubo algún día con menos actividad? ¿Cambian las conclusiones si analizas cada ciudad por separado?`

Etapa 3.4. Los días viernes se escucha más música, debe ser por que es fin de semana, le sigue los lunes y los días miércoles se escucha menos música en estas dos ciudades

Hasta ahora has aprendido a contar entradas agrupándolas por un solo criterio, como la ciudad o el día de la semana. Ahora vamos a dar un paso más: necesitas crear una función que **cuente cuántas canciones se reprodujeron en una ciudad específica durante un día determinado**, combinando ambos criterios de filtrado.

La función se llamará `number_tracks()` y debe aceptar dos parámetros:

- `day`: un día de la semana (por ejemplo, `'Monday'`).
- `city`: el nombre de una ciudad (por ejemplo, `'Springfield'`).

Dentro de la función, deberás aplicar un **filtrado secuencial mediante indexación lógica**: primero tendrás que filtrar el DataFrame por el día, y luego por la ciudad. Una vez que tengas el subconjunto de datos correcto, debes contar cuántas veces aparece un valor en la columna `'user_id'`. Ese número representará el total de canciones reproducidas bajo esos dos criterios.



Etapa 3.5.
1. Declara una función llamada `number_tracks()` con dos parámetros: `day` y `city`.

2. Filtra el DataFrame para conservar solo las filas donde la columna `'day'` sea igual al valor del parámetro `day`.

3. A partir del resultado anterior, filtra nuevamente para conservar solo las filas donde la columna `'city'` sea igual al valor del parámetro `city`.

4. Extrae la columna `'user_id'` del DataFrame filtrado y utiliza el método `.count()` para contar el número de entradas.

5. Guarda ese valor en una variable y **devuélvelo** como resultado de la función.


In [39]:
# Declara la función number_tracks() con dos parámetros: day= y city=.
def number_tracks(day, city):
    # Almacena las filas del DataFrame donde el valor en la columna 'day' es igual al parámetro day=
    df_day = df[df['day'] == day]
    # Filtra las filas donde el valor en la columna 'city' es igual al parámetro city=
    df_city = df_day[df_day['city'] == city]
    # Extrae la columna 'user_id' de la tabla filtrada y aplica el método count()
    entries_number = df_city['user_id'].count()
    # Devuelve el número de valores de la columna 'user_id'
    return entries_number

Etapa 3.6. Llama a `number_tracks()` seis veces, cambiando los valores de los parámetros para que puedas recuperar los datos de ambas ciudades para cada uno de los tres días.

In [40]:
# El número de canciones reproducidas en Springfield el lunes
number_tracks('Monday', 'Springfield')

15740

In [41]:
# El número de canciones reproducidas en Shelbyville el lunes
number_tracks('Monday', 'Shelbyville')

5614

In [42]:
# El número de canciones reproducidas en Springfield el miércoles
number_tracks('Wednesday', 'Springfield')

11056

In [43]:
# El número de canciones reproducidas en Shelbyville el miércoles
number_tracks('Wednesday', 'Shelbyville')

7003

In [44]:
# El número de canciones reproducidas en Springfield el viernes
number_tracks('Friday', 'Springfield')

15945

In [45]:
# El número de canciones reproducidas en Shelbyville el viernes
number_tracks('Friday', 'Shelbyville')

5895

[Volver a Contenidos](#back)

# Conclusiones <a id='end'></a>

## Escribe tus conclusiones finales sobre la hipótesis.
 * ¿Los datos apoyan la idea de que el comportamiento de los usuarios respecto a la música que escuchan varía según la ciudad y el día de la semana?
 * Indica si la hipótesis debe aceptarse o rechazarse, y justifica tu respuesta con base en los resultados obtenidos.

Etapa 4. Si, la actividad de los usuarios y usuarias difiere según el día de la semana y la ciudad. Como se puede observar en la ciudad de Springfield se escucha más música en cualquiera de los 3 días de de la semana (Lunes, miércoles y viernes) en comparación con la ciudad de Shelbyville.
La hipótesis debe aceptarse, dado que, según los resultados obtenidos del análisis pudimos identificar lo siguiente:
* De acuerdo al día de la semana y ciudad varía la actividad de los usuarios con la música.
* En Springfield (42741) se escucha más música que en Shelbyville (18512).
* Los días lunes (15740) y viernes (15945) en la ciudad de Springfield se escucha más música que los miércoles (11056), pero pasa lo contrario en la ciudad de Shelbyville, donde los miércoles(7003) se escucha más música que los Lunes (5614) y Viernes (5895).

### Nota
En los proyectos de investigación reales, la prueba de hipótesis estadística es más precisa y cuantitativa. También ten en cuenta que no siempre se pueden sacar conclusiones sobre una ciudad entera a partir de datos de una sola fuente.

Aprenderás más sobre la prueba de hipótesis en el sprint de análisis estadístico de datos.

[Volver a Contenidos](#back)